### ISSUE TITLE
Tenanted documents storing in non-tenant session

### ISSUE THREAD

**Issue by barclayadam** [JasperFx/marten#1460]
I have a multi-tenanted app that, within our API, has sessions scoped to a particular tenant per request. 

In addition, we have background processing that, by its nature, needs to not be tenant specific.

A few things I've come across which I'm either missing or is perhaps an indication of some changes needed in Marten are below. Sorry it's a bit of a mix of changes, thoughts and potential bugs, just wanted to get it all written down now

**LoadAsync**

`LoadAsync(id)` has no way of indicating "any tenant". This can be worked around using `Query<T>` with `.AnyTenant()`, but is perhaps an area that could be improved for discoverability

**Updates**

It is not possible to easily update documents that have been loaded that are document-specific. In a non-tenant session I load a tenant-specific document (as above), then call `Store` or `Update`. Because of the session-level tenant ID that update fails with a `NonExistentDocumentException`. 

We can workaround by adding some extension methods that can look for tenant-specific documents that we add an interface to, but it would be great if the tracking would identify the tenant a document is loaded from for future updates.

**ITenantSession**

This is a bigger change, and not sure on the exact semantics, but I was thinking it may be a nice idea to have an interface specific to tenant-scoped sessions. In an app that needs both tenant and non-tenant sessions, it would be great to help identify those areas that _must_ be tenant-specific through the type system.

It would also potentially help solidify the API in terms of what is available (i.e. Tenant-specific information not needed in non-tenant specific interface).

**Tenant-scoped session view**

Something I've found could have been useful is opening a tenant-specific session from a non-tenant specific one, sharing the same connection etc. It would work on the same UoW etc. but in essence be a "view" of a single tenant. 

The use case I have is a background job that works for any tenant, but at any one time only processes a single one, and I would like components it uses to be given a tenant-specific session to avoid them accidentally loading / saving documents to the wrong tenant.

**Comment by tonykaralis** [comment]
I have pretty much the same use cases and agree this is an area that could do with some improvement.

On your section about Updates, we suffer from the same problem when performing deletes as well. Would you mind sharing the extension method used to help manage the updates?

**Comment by tonykaralis** [comment]
Linking in #1451 as seems relevant

**Comment by barclayadam** [comment]
@tonykaralis I changed the implementation for updates to be a session listener, which works fine for update, insert and upsert operations. You can see that at https://gist.github.com/barclayadam/f4677e027588b7c9da91a73052281cc6

I've not actually come across the deletion issue and having a very quick glance I'm not sure the best way to solve this is, as a delete appears to not use the same base class (`DocumentStorageOperation`). Looking at the code for `DeleteById` I cannot see an easy way to affect it's SQL generation that would be "safe" (i.e. does not just turn off tenant-specific filters).

**Comment by tonykaralis** [comment]
Thanks @barclayadam I will take a look but I think you are right that wouldn't quite work for deletes.

**Comment by jeremydmiller** [comment]
@barclayadam @tonykaralis Hey, I'd already chatted w/ Tony on #1451. We made some structural changes in the V4 code (master) that should make all of this a lot easier now. But I'm wondering if you could expound on the `ITenantSession` idea?

**Comment by barclayadam** [comment]
@jeremydmiller Hey. The idea was to make it very obvious within the code base whether a session has been created with a tenant or not.

My app is set up to create a tenant-specific document session whenever a controller requests `IDocumentSession` AND I have the current tenant (grabbed from the user context).

This works fine so long as the tenant context is available. If it was possible to request `ITenantDocumentSession` or similar then I would _know_ that the controller *needs* a tenant-based session and could therefore fail the request for one if it is not possible to service.

At the moment the factory has no idea and therefore falls back to the default tenant, which could inadvertently lead to entities being stored with the incorrect tenant information.

I'm not sure if I'm explaining the needs too well TBH :) TL;DR I want code that consumes sessions to be able to _enfore_ the state of a session to have the tenant information through the type system.

**Comment by mysticmind** [comment]
I remember we added `StoreOptions.DefaultTenantUsageEnabled` flag, if it is set to `false` then n any operation with a default tenant would throw `DefaultTenantUsageDisabledException`.  Settings this flag to `false` is equivalent to ensuring a non-default tenant is required. Does this help with your problem at hand?

**Comment by barclayadam** [comment]
@mysticmind It does not help because at the time of creating the session I still do not know whether it's valid to have a default tenant session or not.

The problem is an application, using DI, that has different concerns. As there is only a single type I cannot make the decision in DI where the sessions are opened as to whether the session needs to be restricted or not. It's a per DI dependency request, not an app level thing

**Comment by jeremydmiller** [comment]
@barclayadam "Knowing" that a session is for a specific tenant or not should be a pretty easy thing to do. Seems kind of like an obvious thing to do in retrospect.

"The problem is an application, using DI, that has different concerns. As there is only a single type I cannot make the decision in DI where the sessions are opened as to whether the session needs to be restricted or not. It's a per DI dependency request, not an app level thing" -- that's an "it depends". Depending on how you do multi-tenancy, it should be possible to detect which tenant you need inside a specific scoped container. In that case, you would just let the DI container deal w/ the tenant detection and construction for you. When you use the `AddMarten()` registration in the recent versions, it's doing the DI registration with a Lambda builder anyway, so there's plenty of opportunity (more in V4) to customize how the session is resolved by the current HTTP request or something like a message context in MassTransit or NServiceBus.

The `ITenantDocumentSession` idea would also be relatively easy.

I'm glad I asked, because I was picturing something like:

```
// IDocumentOperations is a cut down version of IDocumentSession that's new in V4
public interface ITenantOperations : IDocumentOperations {
    string TenantId {get;}
}

// which you could get to maybe from:

IDocumentSession.ForTenant(string tenantId) : ITenantOperations


```

with the above just so it was easy to emulate a single named tenant at will no matter how you got the `DocumentSession`.

**Comment by barclayadam** [comment]
@jeremydmiller That is exactly how I am managing to modify the session based on the current HTTP context, through a lambda registration of `IDocumentSession` and a different ambient "current tenant" service that uses HTTP context. That is an easy(ish) problem, already solvable now.

The addition of something like `ITenantDocumentSession` is ideal. Simply an indication that what you are trying to resolve MUST be for a valid tenant and if for whatever reason you do not have the tenant, you can fail to resolve the dependency at that time.

Having `ForTenant` on `IDocumentSession` would also solve a different issue I have in the application where I have an operation that starts out non tenant-specific but then I need to open a tenant-specific session for a small part of that operation. Would the `ITenantOperations` returned from `ForTenant` re-use all the existing connections, unit of work etc. of the parent? Or would it act like opening a new session?

TL;DR; I think what you are proposing is pretty much exactly what I had in mind and would solve my issue nicely

**Comment by jeremydmiller** [comment]
For now, I'd like to say that the new `IDocumentSession.ForTenant()` idea is enough and closing this today with a push coming in a minute.
